ykrt/src/compile/jitc_yk/codegen/x86_64/lsregalloc.rs

```diff
@@ -0,0 +1,1099 @@
+                .gp_reg_states
+                .iter()
+                .position(|x| {
+                    if let RegState::FromInst(y) = x {
+                        *y == iidx
+                    } else {
+                        false
+                    }
+                })
+                .unwrap();
+            let cur_reg = GP_REGS[reg_i];
+            if cur_reg != reg {
+                match size {
+                    1 => dynasm!(asm ; movzx Rq(reg.code()), Rb(cur_reg.code())),
+                    2 => dynasm!(asm ; movzx Rq(reg.code()), Rw(cur_reg.code())),
+                    4 => dynasm!(asm ; mov Rd(reg.code()), Rd(cur_reg.code())),
+                    8 => dynasm!(asm ; mov Rq(reg.code()), Rq(cur_reg.code())),
+                    _ => todo!("{}", size),
+                }
+            }
+            return;
+        }
+        debug_assert_ne!(frame_off, usize::MAX);
+        let off = i32::try_from(frame_off).unwrap();
+        match size {
+            1 => dynasm!(asm ; movzx Rq(reg.code()), BYTE [rbp - off]),
+            2 => dynasm!(asm ; movzx Rq(reg.code()), WORD [rbp - off]),
+            4 => dynasm!(asm ; mov Rd(reg.code()), [rbp - off]),
+            8 => dynasm!(asm ; mov Rq(reg.code()), [rbp - off]),
+            _ => todo!("{}", size),
+        }
+        self.gp_regset.set(reg);
+    }
+
+    /// Load the constant from `cidx` into `reg`.
+    fn load_const_into_gp_reg(&mut self, asm: &mut Assembler, cidx: ConstIdx, reg: Rq) {
+        match self.m.const_(cidx) {
+            Const::Float(_tyidx, _x) => todo!(),
+            Const::Int(tyidx, x) => {
+                let Ty::Integer(width) = self.m.type_(*tyidx) else {
+                    panic!()
+                };
+                // The `as`s are all safe because the IR guarantees that no more than `width` bits
+                // are set in the integer i.e. we are only ever truncating zeros.
+                match width {
+                    1 | 8 => dynasm!(asm; mov Rb(reg.code()), BYTE *x as i8),
+                    16 => dynasm!(asm; mov Rw(reg.code()), WORD *x as i16),
+                    32 => dynasm!(asm; mov Rd(reg.code()), DWORD *x as i32),
+                    64 => dynasm!(asm; mov Rq(reg.code()), QWORD *x as i64),
+                    _ => todo!(),
+                }
+            }
+            Const::Ptr(x) => {
+                dynasm!(asm; mov Rq(reg.code()), QWORD *x as i64)
+            }
+        }
+    }
+
+    /// Get an empty general purpose register, freeing one if necessary. Will not touch any
+    /// registers set in `avoid`.
+    fn get_empty_gp_reg(&mut self, asm: &mut Assembler, iidx: InstIdx, avoid: RegSet<Rq>) -> Rq {
+        match self.gp_regset.find_empty_avoiding(avoid) {
+            Some(reg) => reg,
+            None => match avoid.find_empty() {
+                Some(reg) => match self.gp_reg_states[usize::from(reg.code())] {
+                    RegState::Reserved | RegState::Empty => unreachable!(),
```

Comment:
I assume that the registers R12, RBP, RSP are marked with `Reserved` so they won't be used here?

Reply:
Correct.